lib/recursive-methods.rb

```diff
@@ -1,49 +1,120 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
+
+# factorial(n) Write a method factorial that accepts an integer
+# parameter n and that uses recursion to compute and return the
+# value of n factorial (also known as n!).
+
+#     e.g. fact(4) = 4 * 3 * 2 * 1 = 24
+
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if (n < 0)
+    raise ArgumentError
+  elsif n == 1 || n == 0
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
+
+# reverse(s) Write a method reverse that accepts a string as a
+# parameter and then returns the reverse of the string by
+# reversing all letters and all words in the string.
+
+# e.g. reverse("hello world") will return "dlrow olleh"
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if (s != "")
+    return s[-1] + reverse(s[0...-1])
+  else
+    return ""
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  return reverse_inplace_internal(s, 0, s.length - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+def reverse_inplace_internal(s, start_i, end_i)
+  if (start_i >= end_i)
+    return s
+  else
+    temp = s[start_i]
+    s[start_i] = s[end_i]
+    s[end_i] = temp
+
+    start_i += 1
+    end_i -= 1
+
+    return reverse_inplace_internal(s, start_i, end_i)
+  end
+end
+
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  if n <= 0
+    return 0
+  elsif n == 1
+    return 2
+  else
+    return 2 + bunny(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def nested(s)
```

Comment:
👍 This works, but you have similar time/space issues with the above methods due to creating new arrays.